include/boost/geometry/strategies/cartesian/intersection.hpp

```diff
@@ -176,36 +176,43 @@ struct cartesian_segments
             }
         }
 
-    public :
-        template <typename Point, typename Segment1, typename Segment2>
-        void calculate(Point& point, Segment1 const& a, Segment2 const& b) const
+        bool use_a() const
         {
-            bool use_a = true;
-
-            // Prefer one segment if one is on or near an endpoint
-            bool const a_near_end = robust_ra.near_end();
-            bool const b_near_end = robust_rb.near_end();
-            if (a_near_end && ! b_near_end)
-            {
-                use_a = true;
-            }
-            else if (b_near_end && ! a_near_end)
+            // Prefer the segment closer to the endpoint.
+            // If both are about equally close, then prefer the longer segment
+            // To avoid hard thresholds, behavior is made fluent:
+            // Calculate comparable length indications,
+            // the longer the segment (relatively), the lower the value
+            // such that the shorter lengths are evaluated higher and will
+            // be preferred.
+
+            auto const cla = comparable_length_a();
+            auto const clb = comparable_length_b();
+            auto const clm = std::max(cla, clb);
+            if (clm <= 0)
             {
-                use_a = false;
-            }
-            else
-            {
-                // Prefer shorter segment
-                promoted_type const len_a = comparable_length_a();
-                promoted_type const len_b = comparable_length_b();
-                if (len_b < len_a)
-                {
-                    use_a = false;
-                }
-                // else use_a is true but was already assigned like that
+                return true;
             }
 
-            if (use_a)
+            // Relative comparible length
+            double const rcla = 1.0 - cla / clm;
+            double const rclb = 1.0 - clb / clm;
+
+            double const eva = robust_ra.edge_value();
+            double const evb = robust_rb.edge_value();
+
+            // Multipliers for edgevalue (ev)
+            // and relative comparible length (rcl)
+            constexpr double mev = 5.0;
+            constexpr double mrcl = 1.0;
+            return mev * eva + mrcl * rcla > mev * evb + mrcl * rclb;
```

Comment:
This is the fix. Instead of just checking `near_end` of a and b, and if not select the smallest segment, we now have a balance: near the end gets priority (factor 5) but if they are equal, then the formula will select the smallest segment (as earlier).
Theoretically (mathematically) the result is the same, using either a or b. But in case of numeric precision, they might be just a bit different from each other, and it can (mainly for validity checks) be important which one is chosen. Though there is also randomness involved.

Reply:
Why using factor 5? Are similar values also tested? Maybe add a comment in the code about that choice for future reference.

Reply:
It's basically already commented
```
// Prefer the segment closer to the endpoint.
// If both are about equally close, then prefer the longer segment
// To avoid hard thresholds, behavior is made fluent.
// Calculate comparable length indications,
// the longer the segment (relatively), the lower the value
// such that the shorter lengths are evaluated higher and will
// be preferred.
```
The essention is that the `mev` is much larger than `mrcl` . I'll add a comment.

Reply:
I was testing to try the effect but in the test program I used, there is no difference between `1` and `5`
I'll come back to this.

Reply:
Checking this with other robustness tests, there is something wrong, there are regressions.
Those are not regularly tested (because they are not unit tests, they work in a different way).

I'm working on it and update later - it's not the balance, but it's something in the other changes for `approximately_equals`

Reply:
So the fixes are unrelated now. This PR (as at the start) fixes the balance, and the promotion type. But not yet the other-types which will follow in another commit.